oak-core/src/main/java/org/apache/jackrabbit/oak/query/QueryEngineImpl.java

```diff
@@ -292,7 +290,38 @@ public Result executeQuery(
             }
         }
     }
-    
+
+    @Override
+    public Result executeQuery(
+            String statement, String language, long limit, long offset,
+            Map<String, ? extends PropertyValue> bindings,
+            Map<String, String> mappings) throws ParseException {
+        return executeQuery(statement, language, Optional.of(limit), Optional.of(offset), bindings, mappings);
+
+    }
+
+    private static long getOffset(List<Query> queries, Optional<Long> passedOffset) {
+        if (!passedOffset.isPresent()) {
+            return queries.stream().map(Query::getOffset).filter(Optional::isPresent).map(Optional::get).findFirst()
+                    .orElse(0L);
+        }
+        if (passedOffset.get() < 0) {
+            throw new IllegalArgumentException("Offset may not be negative, is: " + passedOffset.get());
+        }
+        return passedOffset.get();
+    }
+
+    private static long getLimit(List<Query> queries, Optional<Long> passedLimit) {
+        if (!passedLimit.isPresent()) {
+            return queries.stream().map(Query::getLimit).filter(Optional::isPresent).map(Optional::get).findFirst()
```

Comment:
I would check for IllegalArgumentException as well here

Reply:
I don't believe an IllegalArgumentException can be thrown here as getLimit is only returning the previously parsed limit optional